### ISSUE TITLE
[question] can we force a copy/update from a mapped volume re: downloaded files?

### ISSUE THREAD

**Issue by borisivan** [zalando/zalenium#249]
Hi,

I have some tests that will examine a file that is downloaded by the browser. It works, but I need to put delays of up to 10-15 seconds to retrieve small files, maybe because there is some kind of background task in which zalenium/docker coordinates getting those files up from the egalu/selenium container, up to the zalenium container (and then from there, mapped to the host running docker itself).

Bottom line, was hoping there is a way I can 'force' the update. I know when it's downloaded, but I would like to force the overall copy of the file across the various mapped volumes elgalu->zalenium->host, so I can inspect the file without a larger arbitrary delay.

Thanks in advance!

**Comment by borisivan** [comment]
Adding the delay doesn't seem to work. If in sequence, this always works. If tests are in parallel, this doesn't seem to work. Could this be a result of the improvements re: maxTestsPerContainer, in that the container directory is cleaned up between tests, and the file that was there hasn't been moved out to the zalenium master container yet?

We launch zalenium with this argument: -v /tmp/Downloads:/tmp/node/home/seluser/Downloads

If I don't have tests run in the same container, the 5 tests that have the browser save a file pass, i.e. the file makes it all the way up from the child container to zalenium to the host /tmp/Downloads. But if I run the tests in parallel, the browser works, and watching the video I can see that the file is downloaded.  But usually 2 or 3 of those 5 tests will fail, the file is never detected on the host.

All 5 files have completely different names.

**Comment by borisivan** [comment]
The tests are not dependent on each other, they can run in any sequence.

beforeSuite
**
sudo rm -R /tmp/Downloads
start zalenium with the mapped dir: -v /tmp/Downloads:/tmp/node/home/seluser/Downloads
sudo chmod 777 /tmp/Downloads

As noted, right after we start zalenium (containers=6, maxTestsPer=150) I also "sudo chmod 777 /tmp/Downloads". This is immediately after we automate the command to start docker, i.e. 1 second later. The child containers have probably not been started yet, or have just come into existence. I wonder if the timing of the chmod is somehow relevant..

Then ~ 100 tests run across the 6 parallel execution containers. At some point these 5 tests run. Those 5 each initiate a browser, navigate to a page that has a download link, and when they click it, I see chrome's download bar at the bottom of the browser show that the file has been downloaded. It is flashing for a second signifying it's still downloading, then it is solid. 10 seconds later, the test attempts to read the file within /tmpDownloads (it's a .csv) to see it's contents. When fails, it's because can't find the file. When the suite is done, sure enough, only the files for the passed tests are in /tmp/Downloads... (i.e. a subset of the 5). If all tests pass, all 5 are there. In _all_ cases, the video looks perfect.

The test(s) do not delete the file, they just read from it. But the beforeSuite code does the rm -R as described above.

**Comment by borisivan** [comment]
This one is really hurting -- I have many tests commented out because this functionality doesn't seem to work. I just eliminated things running in parallel, and ran the test twice. One time it passed, (the file is made available to the docker host), the other time, no file was made available. I put a 5 minute delay to ensure the containers remained open, browser just sitting there.

the browser clicks a link that downloads a small file. Sometimes that makes it up to the host directory, sometimes it doesn't. And when it doesn't -- it never does -- waiting longer doesn't fix it.

**Comment by borisivan** [comment]
Was troubleshooting and found something pretty relevant: sometimes, some of the child containers do not get the mapping set up in the first place!!

root@ubuntu1604:/# docker container inspect zal_40000 | grep -i downloads
                "/tmp/Downloads:/home/seluser/Downloads",
                "Source": "/tmp/Downloads",
                "Destination": "/home/seluser/Downloads",
**root@ubuntu1604:/# docker container inspect zal_40001 | grep -i downloads**
root@ubuntu1604:/# docker container inspect zal_40002 | grep -i downloads
                "/tmp/Downloads:/home/seluser/Downloads",
                "Source": "/tmp/Downloads",
                "Destination": "/home/seluser/Downloads",

in this case, I had 6 containers. 5 had the mapping, 1 did not. In the case above, it was zal_40001. All 6 seem to be otherwise functioning, happy Zalenium containers, and they're all running my tests in parallel. But only 5 have that directory mapped.

root@ubuntu1604:/# docker ps -a
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS                                                NAMES
1b6dd08eeb99        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40005->40005/tcp, 0.0.0.0:50005->50005/tcp   zal_40005
c0d7a0fcd834        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40004->40004/tcp, 0.0.0.0:50004->50004/tcp   zal_40004
7b690d82b344        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40003->40003/tcp, 0.0.0.0:50003->50003/tcp   zal_40003
24790837bdc2        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40002->40002/tcp, 0.0.0.0:50002->50002/tcp   zal_40002
ddbbb73af3ec        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40001->40001/tcp, 0.0.0.0:50001->50001/tcp   zal_40001
47aec5e71039        elgalu/selenium:latest   "entry.sh"               4 minutes ago       Up 4 minutes        0.0.0.0:40000->40000/tcp, 0.0.0.0:50000->50000/tcp   zal-ci-mjl_40000
4d556d6bcc95        dosel/zalenium           "entry.sh start --..."   4 minutes ago       Up 4 minutes        0.0.0.0:4444->4444/tcp, 4445/tcp                     zal-ci-mjl
root@ubuntu1604:/#

A few minutes earlier, all 6 had the mapping, and I assume the tests would have passed if those specific tests weren't still commented out in my suite.

When I posted earlier that I tried running without in parallel, I had 2 child containers created (i.e. the zalenium default). One had the mapping, one didn't.

I assume that this completely explains the intermittent results I was seeing -- when my test that utilizes that mapping happens to run in a container that doesn't have the mapping set up, obviously that's the one that fails.

**Comment by borisivan** [comment]
... and the mapping never fails for the video directory. Just this directory. And only occasionally, yet zalenium is being started the same way, every time. (and even when it fails, it only fails on a subset of those child containers). I posted a generic question along these lines [on stackoverflow](https://stackoverflow.com/questions/46792080/mapping-multiple-docker-containers-to-same-host-directory-fails-occasionally) -- but no responses yet, so it seems doubtful that this is a widely known 'docker bug'.

This is causing us to leave portions of our already automated tests disabled, since they fail when this mapping doesn't exist.

**Comment by diemol** [comment]
Sorry, have not had lots of time to check into the open issues, but not we are trying to clean up a bit.

So, there is no issue with the videos folder because this one is only mounted between the host and the Zalenium container, not across the containers.

Now, about the mounted volume across all child containers, I am kind of stuck here because I started 6 containers on my machine and I don't see the behaviour. But I have a Mac, so perhaps this makes the difference in this issue. In addition, I have not seen this reported somewhere else, so maybe it is something with your machine or your use case is not common.

Anyway, I would really like to dig deep into it, is there any way that you could provide me with the test setup and the test script your run (or a similar one), so I can try to reproduce it?

P.S. Have you tested in different machines? E.g. a Mac or a someone else's ubuntu?

**Comment by borisivan** [comment]
I see the exact same problem on 2 different machines, different OS (the Ubuntu node above, as well as SLES 12).

I'll get the exact command line that I'm automating to launch Zalenium and update the comments. That's the only thing needed -- I don't even need to run a test -- when I start Zalenium configured to start the 6 child containers, it's easy to know if the problem will happen, just by looking at the docker container inspect command (in the comment above -- see example) when things have finished loading. Most (sometimes all, which is good) containers will have the mapping, and sometimes 1 or 2 won't. Those are the ones that do run the test, but I won't get the downloaded file from the browser, because the mapping doesn't even exist. But other than that it is a healthy 'child' container capable of running tests.

When it occurs, waiting doesn't help, that mapping never exists.

**Comment by diemol** [comment]
OK, then I'll wait for your input so I try to reproduce it in different environments.

**Comment by borisivan** [comment]
docker run -tid --name=xxxl -p 4444:4444 -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker -v /tmp/videos:/home/seluser/videos -v /tmp/Downloads:/tmp/node/home/seluser/Downloads --privileged dosel/zalenium start --chromeContainers 6 --screenWidth 1920 --screenHeight 1080 --firefoxContainers 0 --videoRecordingEnabled true --maxTestSessions 150

**Comment by diemol** [comment]
Hi @borisivan,

Yes, I think #296 might be the reason. It would make sense. Lets work on solving that one and then we check back on this one, I think this one will get solved as well.

**Comment by diemol** [comment]
@borisivan #296 was fixed and released, can you please try again?

**Comment by dgracehmh** [comment]
@borisivan Did this fix the issue for you? I am seeing a very similar issue when I am running lots of tests and trying to check the downloaded file. Even with long waits for the file to be downloaded I cannot see/check it. Just started now to try debug.

**Comment by diemol** [comment]
@dgracehmh if you are using the latest release and you are running into that issue, is there a way that you can extract a test case for me to reproduce it and fix it?

**Comment by dgracehmh** [comment]
False alarm @diemol , i've tracked it down to using headless chrome : -https://bugs.chromium.org/p/chromedriver/issues/detail?id=1973&q=headless&colspec=ID%20Status%20Pri%20Owner%20Summary

**Comment by borisivan** [comment]
Thanks @diemol , sorry I didn't update the report -- got caught up in a few things. Yep, things are looking good!
